### ISSUE TITLE
Excited...

### ISSUE THREAD

**Issue by KipourosV** [maksis/airdcpp-release-validator#1]
I have been working with a friend on a couple of scripts for sfv and dupes checking.

Take a look...
https://github.com/KipourosV/MusicTools
Fairly simple python scripts.

For the SFV check part, my idea goes like this:
There are 4 states for the release(folder).

* New (not checked)
* Complete (all sfv veryfied ok)
* Incomplete (some sfv verified, some missing files) and
* Missing (no sfv file in folder).

First argument is the root path that will be recursively scanned.
When you run the script with no second argument, it just reports the subfolders(releases) state.

Second argument is the state that you want checked.
When you run the script with second argument (for example new)
the script sfv checks all the folders that are "new", then writes a state file in
each checked directory, 0byte size and named after the resulting state.

The second script recursivelly parses the given folder and does a fuzzy search on the folder names
in an attempt to locate from exact duplicates to similar release names (ex. from different groups).
It also has a similarity ratio that you can tweak.

The idea behind all this is to automate:
1) mass (auto)-redownloading incomplete releases
2) mass deleting garbage folders
3) compare releases from different groups (and keep the best quality rls based on size)
4) (auto)cleaning the share of duplicates
5) being able to initiate a targeted sfv recheck

I would really like to help out with this extension in any way I can.
I need some guidance, I m not advanced in programming but I learn quickly.

How do I work with your code here?
Where do I need to place what in order to start looking at how the extension works and what it does?

**Comment by maksis** [comment]
Oh, share scanner in AirDC++ was one of the first real application features that I've ever developed. That feature is old and well tested but it shouldn't really be part of the core application so I'd like to move all the validation functionality into this extension and make things more modular (just look at the actual validation function: https://github.com/airdcpp/airgit/blob/master/airdcpp/airdcpp/modules/ShareScannerManager.cpp#L331-L578).

Auto download of missing files is certainly a popular request. One way to allow that would be to make it possible for extensions to communicate with each others, so that you could have another extension listening for "directory scan failed" events from this extension and performing actions based on them. My plan is to keep this extension focused on the directory content structure validation and let other extensions handle dupe checks, CRC validations and other features (those are quite obvious starting points for extension development as well).

I'll post an update once extension support has been enabled in the develop branch as you can't do much before that.

**Comment by KipourosV** [comment]
I will do some reading on nodejs as am fairly new to Javascript...
Really looking forward to this!

**Comment by maksis** [comment]
Extensions are now supported in the develop version. Extensions written in other languages are supported as well (I need to document that at some point) but yes, Node will be the primary one as it's going to be shipped with the Windows version.

**Comment by maksis** [comment]
Seems like there are some issues with older versions of node but at least the latest ones (v6 and v7) seem to be working

**Comment by maksis** [comment]
It was just memory corruption caused by the app itself so everything should be fine now :p

**Comment by maksis** [comment]
@KipourosV  Do you find the [starter project](https://github.com/airdcpp-web/airdcpp-create-extension) and examples to be useful?

Some feature ideas for this extension: 

- Validations for sample directories and zip releases
- Allow skipping validation for files/directories that are ignored from share
- Allow configuring custom directories that shouldn't be validatated

**Comment by KipourosV** [comment]
I haven't had time to fiddle with it yet.
I'm in the middle of recovering a degraded ceph storage cluster and a pregnant wife.

I'll give you an update this weekend...

**Comment by maksis** [comment]
Have you abandoned this project? :p

**Comment by KipourosV** [comment]
No and yes, I've been using the scripts to do the job since I don't have
enough time to dive into nodejs.

Any news from your side? Any ready code that I could look at to get an
example?

**Comment by maksis** [comment]
The starter project might be helpful: https://github.com/airdcpp-web/airdcpp-create-extension

I've also written a really simple standalone Python extension: https://github.com/airdcpp-web/airdcpp-example-python-extension

**Comment by KipourosV** [comment]
Also another nice idea would be to have a window of it's own with buttons to initiate the release validator for example, and maybe some sort of activity display (is it running? what is it doing?). Instead of flooding the main log, it could have a window of it's own...

**Comment by KipourosV** [comment]
Also, the results of the release validator (for example the folders that are missing some files) could be copy pasted to the auto downloader extension that I mentioned earlier. So you can get the files you are missing automatically. Also a way to auto download them at the duplicate location would be sweet!

**Comment by maksis** [comment]
> Also, the results of the release validator (for example the folders that are missing some files) could be copy pasted to the auto downloader extension that I mentioned earlier. So you can get the files you are missing automatically. Also a way to auto download them at the duplicate location would be sweet!

Yes, an extension like that should be doable already (add a setting text area and subscribe for setting updates)

**Comment by maksis** [comment]
There's also an API method that will return dupe paths in share based on the release name: https://airdcpp.docs.apiary.io/#reference/share/generic-methods/get-dupe-paths

**Comment by KipourosV** [comment]
Can you show me an example of how you initiate a search through an api
call? And also how to initiate a download?

**Comment by KipourosV** [comment]
I'm gonna start working on this, waiting on a pro progamer friend for some
help.
I have so many ideas for this...
Most hubs have release announcements, on main chat or via an msg.
The extension should be able to parse and auto download, according to your
filter.
For example... This is the announcement:

[[[Pr3.anounce added to the category [Psychedelic-FLAC] the following
Release: VA_-_Mush-A-Holic_(Compiled_By_DJ_Mush)-(WMCD014)-CD-FLAC-2007-VL

So in the extensions settings you can have filters that you can
enable/disable/add/remove like :

Filter name: PsyFlac (friendly name)
Filter value: [[[Pr3.anounce added to the category [Psychedelic-FLAC] the
following Release:

I'm getting to work!

**Comment by maksis** [comment]
Feel free to open a new issue (or visit the dev hub) if you have additional questions
